[minosx31] iP

minosx31 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [(Source)](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is __FREE!__

Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

***

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
  public static void main(String[] args) {
    Application.launch(MainApp.class, args);
  }
}
```

Commit:
Add Gradle support

minosx31 committed:
Increment Level-1

minosx31 committed:
Increment Level-2

minosx31 committed:
Increment Level-3

minosx31 committed:
Increment Level-4

minosx31 committed:
Increment A-TextUiTesting

minosx31 committed:
Increment Level-5

minosx31 committed:
Increment Level-6

minosx31 committed:
Increment A-Enums

LiuMC-SG commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+
+    private static String numOfTasks(ArrayList<Task> tasks) {
+        return tasks.size() == 1 ? " task " : " tasks ";
+    }
```
The method name could perhaps be changed as it is misleading. numOfTasks would normally imply returning the number of tasks the list contain. However, you are just getting the string of whether it task should be plural or not. Method names should also start with a verb instead.

LiuMC-SG commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
                 + "|____/ \\__,_|_|\\_\\___|\n";
-        System.out.println("Hello from\n" + logo);
+
+        System.out.println("Hello I'm Duke!\n" + "How can I help you?");
```
Maybe you can take out the string in your println to avoid "magic" strings. Such as welcomeMsg = "...". There are several instances later on as well that you can take not of.

LiuMC-SG commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+    @Override
+    public String toString() {
+        return "[D]" + "[" + super.getStatusIcon() + "] " + super.getDescription() + " (by: " + this.doBy + ")";
+    }
```
I feel like there is no need to use `super.getStatusIcon` and `super.getDescription` since your Deadline already extends from Tasks. Hence, these methods already exist within the Deadline class. Furthermore, the use of this should be removed if there is no multiple instances of it. This is a suggestion.
```suggestion
    public String toString() {
        return String.format("[D] [%s] %s (by: %s)", getStatusIcon(), getDescription(), doBy);
    }
```
The other classes that extend from Task have similar problem and can be dealt with respectively as well.

LiuMC-SG commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+                    System.out.println("OK, I've marked this task as not done yet:");
+                    System.out.println("[" + currentTask.getStatusIcon() + "] " + currentTask.getDescription());
+                } else if (command.equals(commands.TODO.toString())) {
+                    if (remainingInput == null) {
+                        throw new IncorrectInputException("Enter a description!");
+                    }
```
I like that you did this so that it is easier to see the "happy path" that the code would traverse through.

LiuMC-SG commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+                    break;
+                } else if (command.equals(commands.LIST.toString())) {
+                    System.out.println("Here are the tasks in your list");
+                    if (tasks.size() != 0) {
+                        for (int i = 0; i < tasks.size(); i++) {
+                            System.out.println((i + 1) + "." + tasks.get(i).toString());
+                        }
+                    }
```
Perhaps the if condition can be taken out so that it is easier to see the "happy path" and reduce indentation of code.
```suggestion
                    if (tasks.size() == 0) {
                        continue;
                    }
                    for (int i = 0; i < tasks.size(); i++) {
                        System.out.println((i + 1) + "." + tasks.get(i).toString());
                    }
```

LiuMC-SG reviewed (commented):
Overall okay so far, just a few key points to take note of to change. Some of the changes such as simplifying the long if-else/switch case would come naturally after the next few levels.
Keep up the good work! :D

venuslimm commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+        }
+    }
+
+    private static String numOfTasks(ArrayList<Task> tasks) {
```
I agree with @LiuMC-SG, perhaps this can be a suggestion?
```suggestion
    private static String getTaskString(ArrayList<Task> tasks) {
```

venuslimm commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+
     public static void main(String[] args) {
+        ArrayList<Task> tasks = new ArrayList<>();
+        Scanner sc = new Scanner(System.in );
```
Should there be no extra space after `System.in`?
```suggestion
        Scanner sc = new Scanner(System.in);
```

venuslimm commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,13 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + "[" + super.getStatusIcon() + "] " + super.getDescription() + " (from: " + this.period[0] + " to: " + this.period[1] + ")";
```
Would it be better to use line break as the character count exceeds 120?

venuslimm commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,121 @@
+                    System.out.println(newTask);
+                    System.out.println("Now you have " + tasks.size() + Duke.numOfTasks(tasks) + "in the list.");
+                } else if (command.equals(commands.DEADLINE.toString())) {
+                    if (remainingInput == null) {
+                        throw new IncorrectInputException("Enter a description with a deadline!");
+                    }
+                    String[] deadlineDetails = remainingInput.split(" /by ", 2);
+                    String deadlineDescription = deadlineDetails[0];
+                    String deadline = deadlineDetails.length > 0 ? deadlineDetails[1] : null;
+
+                    if (deadline == null) {
+                        throw new IncorrectInputException("Enter a valid deadline!");
+                    }
+                    Task newTask = new Deadline(deadlineDescription, deadline);
+                    tasks.add(newTask);
+                    System.out.println("Got it. I've added this task:");
+                    System.out.println(newTask);
+                    System.out.println("Now you have " + tasks.size() + Duke.numOfTasks(tasks) + "in the list.");
```
I like how the different logical units are separated by a line! The same goes for the rest of this if statement.

venuslimm reviewed (commented):
LGTM! I have commented on a few suggestions but generally, I think your code is easy to read! I like how you extracted the Task class very neatly into the 3 other classes.

minosx31 committed:
Increment Level-7

minosx31 committed:
Increment Level-8

minosx31 committed:
Merge branch 'branch-Level-7'

minosx31 committed:
Merge branch 'branch-Level-8'

minosx31 committed:
Merge branch-Level-7 and branch-Level-8

minosx31 committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

minosx31 committed:
Increment A-MoreOOP

minosx31 committed:
Increment A-Packages

minosx31 committed:
Increment A-Gradle

minosx31 committed:
Increment A-JUnit

minosx31 committed:
Increment A-Jar

minosx31 committed:
Added JavaDoc comments

minosx31 committed:
Follow coding standard

minosx31 committed:
Increment Level-9

minosx31 committed:
Merge branch 'branch-A-JavaDoc'

minosx31 committed:
Resolve merge conflict with branch-A-CodingStandard

minosx31 committed:
Merge branch 'branch-Level-9'

minosx31 committed:
Increment A-CheckStyle

minosx31 committed:
Increment Level-10

minosx31 committed:
Merge branch 'branch-A-CheckStyle'

minosx31 committed:
Merge branch 'branch-Level-10'

minosx31 committed:
Add assertions

The current Command classes that are responsible for handling commands
that involve creating a new task, assumes that a Task object will be
created upon execution. These classes are:
* DeadlineCommand
* EventCommand
* TodoCommand

These assumptions will have to hold in order for these classes to work
as intended. However, if they do not hold, it will result in errors in
other areas of the code.

Having assertions will ensure that the original assumption holds in the
respective Command classes.

minosx31 committed:
Improve code quality

Ensure that the code follows the provided guidelines so that it
remains readable and is of high quality.

minosx31 committed:
Merge pull request #2 from minosx31/branch-A-Assertions

Add assertions

minosx31 committed:
Merge branch 'master' into branch-A-CodeQuality

minosx31 committed:
Merge pull request #3 from minosx31/branch-A-CodeQuality

Improve code quality

minosx31 committed:
Add gradle.yml

minosx31 committed:
Add C-Help extension

minosx31 committed:
Merge branch 'branch-C-Help'

minosx31 committed:
Improve GUI

minosx31 committed:
Merge branch 'branch-A-BetterGui'

minosx31 committed:
Update Duke's responses

minosx31 committed:
Merge branch 'branch-A-Personality'

minosx31 committed:
Update ByeCommand

minosx31 committed:
Add more error handling

MarkCommand does not handle the scenario where a task number does not
exist in the TaskList. Todo, Event and Deadline also does not handle
scenarios where task details are empty or invalid.

When a user enters an input that belongs to such scenarios, they may
encounter runtime errors which are undesirable.

The respective classes can now handle errorneous inputs such as:
* marking/unmarking a task number that does not exist
* wrong command format
* empty task descriptons

minosx31 committed:
Update  of Event and Deadline

minosx31 committed:
Merge branch 'branch-A-MoreErrorHandling'

minosx31 committed:
Update GUI

minosx31 committed:
Remove redundant code

minosx31 committed:
Fix typo

minosx31 committed:
Update README.md and add Ui.png

minosx31 committed:
Edit build.gradle file

minosx31 committed:
Update data file

minosx31 committed:
Update mainClassName

minosx31 committed:
Fix test cases

minosx31 committed:
Fix checkstyle errors
